common/src/api/external/mod.rs

```diff
@@ -1832,11 +1833,108 @@ pub struct VpcFirewallRuleFilter {
 
 /// The protocols that may be specified in a firewall rule's filter
 #[derive(Clone, Copy, Debug, PartialEq, Deserialize, Serialize, JsonSchema)]
-#[serde(rename_all = "UPPERCASE")]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "type", content = "value")]
 pub enum VpcFirewallRuleProtocol {
     Tcp,
     Udp,
-    Icmp,
+    Icmp(Option<VpcFirewallIcmpFilter>),
+    // TODO: IPv6 not supported by instances.
+    // Icmpv6(Option<VpcFirewallIcmpFilter>),
+    // TODO: OPTE does not yet permit further L4 protocols. (opte#609)
+    // Other(u16),
+}
+
+impl FromStr for VpcFirewallRuleProtocol {
+    type Err = String;
```

Comment:
i wonder if this (and all the other `FromStr` impls) ought to return [`omicron_common::api::external::Error::InvalidValue`](https://github.com/oxidecomputer/omicron/blob/5816423e9ce776b80d0fd180d9e9c4fb6ad8c4e2/common/src/api/external/error.rs#L56-L57), rather than `String`/`&'static str`s? It's not hugely important, but gets us more consistent formatting with other API errors, and a place to indicate what value was invalid...

Reply:
I've given this a go for the firewall-related types and I think it works well enough. I'll open an issue on the other `FromStr`s in here.

Reply:
Opened https://github.com/oxidecomputer/omicron/issues/8238.